cl/search/docket_number_cleaner.py

```diff
@@ -176,3 +241,206 @@ def clean_docket_number_raw(
+            response_model=CleanDocketNumber,
+            temperature=0.1,
+            max_completion_tokens=1000,
+            api_key=env("DOCKET_NUMBER_CLEANING_API_KEY"),
+        )
+    except ValidationError as e:
+        logger.error(
+            "LLM - Response validation error for batch=%s",
+            f"{batch}",
+            extra={
+                "validation_errors": e.errors(),
+                "fingerprint": ["llm-docket-number-validation-error"],
+            },
+        )
+        return
+    except Exception as e:
+        # Only expect to get instructor exceptions here to track them
+        capture_exception(e)
+        raise
+
+    if not isinstance(llm_response, CleanDocketNumber):
+        # Added this to avoid mypy errors
+        logger.error("LLM - Invalid response type: %s", type(llm_response))
+        return
+
+    records = {}
+    for item in llm_response.docket_numbers:
+        docket_id = item.unique_id
+        docket_num = [s.upper() for s in sorted(list(set(item.cleaned_nums)))]
+        records[int(docket_id)] = "; ".join(docket_num)
+
+    return records
+
+
+def process_llm_batches(
+    llm_batches: list[dict[int, str]],
+    system_prompt: str,
+    model_id: str,
+    retry: int,
+    max_retries: int,
+    batch_size: int,
+    all_cleaned: dict[int, str],
+) -> dict[int, str]:
+    """
+    Processes batches of records using a language model (LLM), recursively retry with smaller batch sizes for failed extractions.
+
+    :param llm_batches: The batch of records to process. Each item is a dictionary with the docket_id (int) as the key and the raw_docket_number (str) as the value.
+    :param system_prompt: The system prompt to provide to the LLM.
+    :param model_id: Identifier for the LLM model to use.
+    :param retry: Current retry attempt count.
+    :param max_retries: Maximum number of allowed retries for failed batches.
+    :param batch_size: Number of records to process in each model call.
+    :param all_cleaned: Accumulator dictionary for all processed records.
+    :return: A dictionary mapping record IDs to their cleaned/extracted values, including any unprocessed records after max retries.
+    """
+    logger.info(f"---Processing {len(llm_batches)} records with {model_id}---")
+    batches = [
+        llm_batches[i : i + batch_size]
+        for i in range(0, len(llm_batches), batch_size)
+    ]
+    with ThreadPoolExecutor(
+        max_workers=settings.DOCKET_NUMBER_CLEANING_MAX_WORKERS
+    ) as executor:
+        future_to_batch = {
+            executor.submit(
+                extract_with_llm, batch, system_prompt, model_id
+            ): batch
+            for batch in batches
+        }
+        for future in as_completed(future_to_batch):
+            parsed_output = future.result()
+            all_cleaned.update(parsed_output or {})
+    # Check for any still-unprocessed records (e.g., if LLM failed to extract)
+    processed_ids = [k for k in all_cleaned.keys()]
```

Comment:
We can make this a SET for a faster membership check.

```suggestion
    processed_ids = set(all_cleaned.keys())
```